File: Kernel Simulator/Misc/Screensaver/Displays/Lighter.vb

'    Kernel Simulator  Copyright (C) 2018-2022  Aptivi
'
'    This file is part of Kernel Simulator
'
'    Kernel Simulator is free software: you can redistribute it and/or modify
'    it under the terms of the GNU General Public License as published by
'    the Free Software Foundation, either version 3 of the License, or
'    (at your option) any later version.
'
'    Kernel Simulator is distributed in the hope that it will be useful,
'    but WITHOUT ANY WARRANTY; without even the implied warranty of
'    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
'    GNU General Public License for more details.
'
'    You should have received a copy of the GNU General Public License
'    along with this program.  If not, see <https://www.gnu.org/licenses/>.

Namespace Misc.Screensaver.Displays
    Public Module LighterSettings

        Private _lighter255Colors As Boolean
        Private _lighterTrueColor As Boolean = True
        Private _lighterDelay As Integer = 100
        Private _lighterMaxPositions As Integer = 10
        Private _lighterBackgroundColor As String = New Color(ConsoleColor.Black).PlainSequence
        Private _lighterMinimumRedColorLevel As Integer = 0
        Private _lighterMinimumGreenColorLevel As Integer = 0
        Private _lighterMinimumBlueColorLevel As Integer = 0
        Private _lighterMinimumColorLevel As Integer = 0
        Private _lighterMaximumRedColorLevel As Integer = 255
        Private _lighterMaximumGreenColorLevel As Integer = 255
        Private _lighterMaximumBlueColorLevel As Integer = 255
        Private _lighterMaximumColorLevel As Integer = 255

        ''' <summary>
        ''' [Lighter] Enable 255 color support. Has a higher priority than 16 color support.
        ''' </summary>
        Public Property Lighter255Colors As Boolean
            Get
                Return _lighter255Colors
            End Get
            Set(value As Boolean)
                _lighter255Colors = value
            End Set
        End Property
        ''' <summary>
        ''' [Lighter] Enable truecolor support. Has a higher priority than 255 color support.
        ''' </summary>
        Public Property LighterTrueColor As Boolean
            Get
                Return _lighterTrueColor
            End Get
            Set(value As Boolean)
                _lighterTrueColor = value
            End Set
        End Property
        ''' <summary>
        ''' [Lighter] How many milliseconds to wait before making the next write?
        ''' </summary>
        Public Property LighterDelay As Integer
            Get
                Return _lighterDelay
            End Get
            Set(value As Integer)
                _lighterDelay = value
            End Set
        End Property
        ''' <summary>
        ''' [Lighter] How many positions to write before starting to blacken them?
        ''' </summary>
        Public Property LighterMaxPositions As Integer
            Get
                Return _lighterMaxPositions
            End Get
            Set(value As Integer)
                _lighterMaxPositions = value
            End Set
        End Property
        ''' <summary>
        ''' [Lighter] Screensaver background color
        ''' </summary>
        Public Property LighterBackgroundColor As String
            Get
                Return _lighterBackgroundColor
            End Get
            Set(value As String)
                _lighterBackgroundColor = value
            End Set
        End Property
        ''' <summary>
        ''' [Lighter] The minimum red color level (true color)
        ''' </summary>
        Public Property LighterMinimumRedColorLevel As Integer
            Get
                Return _lighterMinimumRedColorLevel
            End Get
            Set(value As Integer)
                _lighterMinimumRedColorLevel = value
            End Set
        End Property
        ''' <summary>
        ''' [Lighter] The minimum green color level (true color)
        ''' </summary>
        Public Property LighterMinimumGreenColorLevel As Integer
            Get
                Return _lighterMinimumGreenColorLevel
            End Get
            Set(value As Integer)
                _lighterMinimumGreenColorLevel = value
            End Set
        End Property
        ''' <summary>
        ''' [Lighter] The minimum blue color level (true color)
        ''' </summary>
        Public Property LighterMinimumBlueColorLevel As Integer
            Get
                Return _lighterMinimumBlueColorLevel
            End Get
            Set(value As Integer)
                _lighterMinimumBlueColorLevel = value
            End Set
        End Property
        ''' <summary>
        ''' [Lighter] The minimum color level (255 colors or 16 colors)
        ''' </summary>
        Public Property LighterMinimumColorLevel As Integer
            Get
                Return _lighterMinimumColorLevel
            End Get
            Set(value As Integer)
                _lighterMinimumColorLevel = value
            End Set
        End Property
        ''' <summary>
        ''' [Lighter] The maximum red color level (true color)
        ''' </summary>
        Public Property LighterMaximumRedColorLevel As Integer
            Get
                Return _lighterMaximumRedColorLevel
            End Get
            Set(value As Integer)
                _lighterMaximumRedColorLevel = value
            End Set
        End Property
        ''' <summary>
        ''' [Lighter] The maximum green color level (true color)
        ''' </summary>
        Public Property LighterMaximumGreenColorLevel As Integer
            Get
                Return _lighterMaximumGreenColorLevel
            End Get
            Set(value As Integer)
                _lighterMaximumGreenColorLevel = value
            End Set
        End Property
        ''' <summary>
        ''' [Lighter] The maximum blue color level (true color)
        ''' </summary>
        Public Property LighterMaximumBlueColorLevel As Integer
            Get
                Return _lighterMaximumBlueColorLevel
            End Get
            Set(value As Integer)
                _lighterMaximumBlueColorLevel = value
            End Set
        End Property
        ''' <summary>
        ''' [Lighter] The maximum color level (255 colors or 16 colors)
        ''' </summary>
        Public Property LighterMaximumColorLevel As Integer
            Get
                Return _lighterMaximumColorLevel
            End Get
            Set(value As Integer)
                _lighterMaximumColorLevel = value
            End Set
        End Property

    End Module
    Public Class LighterDisplay
        Inherits BaseScreensaver
        Implements IScreensaver

        Private RandomDriver As Random
        Private CurrentWindowWidth As Integer
        Private CurrentWindowHeight As Integer
        Private ResizeSyncing As Boolean
        Private ReadOnly CoveredPositions As New List(Of Tuple(Of Integer, Integer))

        Public Overrides Property ScreensaverName As String = "Lighter" Implements IScreensaver.ScreensaverName

        Public Overrides Property ScreensaverSettings As Dictionary(Of String, Object) Implements IScreensaver.ScreensaverSettings

        Public Overrides Sub ScreensaverPreparation() Implements IScreensaver.ScreensaverPreparation
            'Variable preparations
            RandomDriver = New Random
            CurrentWindowWidth = ConsoleWrapper.WindowWidth
            CurrentWindowHeight = ConsoleWrapper.WindowHeight
            SetConsoleColor(New Color(LighterBackgroundColor), True)
            ConsoleWrapper.Clear()
            Wdbg(DebugLevel.I, "Console geometry: {0}x{1}", ConsoleWrapper.WindowWidth, ConsoleWrapper.WindowHeight)
        End Sub

        Public Overrides Sub ScreensaverLogic() Implements IScreensaver.ScreensaverLogic
            ConsoleWrapper.CursorVisible = False

            'Select a position
            Dim Left As Integer = RandomDriver.Next(ConsoleWrapper.WindowWidth)
            Dim Top As Integer = RandomDriver.Next(ConsoleWrapper.WindowHeight)
            WdbgConditional(ScreensaverDebug, DebugLevel.I, "Selected left and top: {0}, {1}", Left, Top)
            ConsoleWrapper.SetCursorPosition(Left, Top)
            If Not CoveredPositions.Any(Function(t) t.Item1 = Left And t.Item2 = Top) Then
                WdbgConditional(ScreensaverDebug, DebugLevel.I, "Covering position...")
                CoveredPositions.Add(New Tuple(Of Integer, Integer)(Left, Top))
                WdbgConditional(ScreensaverDebug, DebugLevel.I, "Position covered. Covered positions: {0}", CoveredPositions.Count)
            End If

            'Select a color and write the space
            If LighterTrueColor Then
                Dim RedColorNum As Integer = RandomDriver.Next(LighterMinimumRedColorLevel, LighterMaximumRedColorLevel)
                Dim GreenColorNum As Integer = RandomDriver.Next(LighterMinimumGreenColorLevel, LighterMaximumGreenColorLevel)
                Dim BlueColorNum As Integer = RandomDriver.Next(LighterMinimumBlueColorLevel, LighterMaximumBlueColorLevel)
                WdbgConditional(ScreensaverDebug, DebugLevel.I, "Got color (R;G;B: {0};{1};{2})", RedColorNum, GreenColorNum, BlueColorNum)
                Dim ColorStorage As New Color(RedColorNum, GreenColorNum, BlueColorNum)
                If CurrentWindowHeight <> ConsoleWrapper.WindowHeight Or CurrentWindowWidth <> ConsoleWrapper.WindowWidth Then ResizeSyncing = True
                If Not ResizeSyncing Then
                    SetConsoleColor(ColorStorage, True)
                    WritePlain(" ", False)
                Else
                    WdbgConditional(ScreensaverDebug, DebugLevel.W, "Resize-syncing. Clearing covered positions...")
                    CoveredPositions.Clear()
                End If
            ElseIf Lighter255Colors Then
                Dim ColorNum As Integer = RandomDriver.Next(LighterMinimumColorLevel, LighterMaximumColorLevel)
                WdbgConditional(ScreensaverDebug, DebugLevel.I, "Got color ({0})", ColorNum)
                If CurrentWindowHeight <> ConsoleWrapper.WindowHeight Or CurrentWindowWidth <> ConsoleWrapper.WindowWidth Then ResizeSyncing = True
                If Not ResizeSyncing Then
                    SetConsoleColor(New Color(ColorNum), True)
                    WritePlain(" ", False)
                Else
                    WdbgConditional(ScreensaverDebug, DebugLevel.W, "Resize-syncing. Clearing covered positions...")
                    CoveredPositions.Clear()
                End If
            Else
                If CurrentWindowHeight <> ConsoleWrapper.WindowHeight Or CurrentWindowWidth <> ConsoleWrapper.WindowWidth Then ResizeSyncing = True
                If Not ResizeSyncing Then
                    Console.BackgroundColor = colors(RandomDriver.Next(LighterMinimumColorLevel, LighterMaximumColorLevel))
                    WdbgConditional(ScreensaverDebug, DebugLevel.I, "Got color ({0})", Console.BackgroundColor)
                    WritePlain(" ", False)
                Else
                    WdbgConditional(ScreensaverDebug, DebugLevel.W, "Resize-syncing. Clearing covered positions...")
                    CoveredPositions.Clear()
                End If
            End If

            'Simulate a trail effect
            If CoveredPositions.Count = LighterMaxPositions Then
                WdbgConditional(ScreensaverDebug, DebugLevel.I, "Covered positions exceeded max positions of {0}", LighterMaxPositions)
                Dim WipeLeft As Integer = CoveredPositions(0).ToString.Substring(0, CoveredPositions(0).ToString.IndexOf(";"))
                Dim WipeTop As Integer = CoveredPositions(0).ToString.Substring(CoveredPositions(0).ToString.IndexOf(";") + 1)
                WdbgConditional(ScreensaverDebug, DebugLevel.I, "Wiping in {0}, {1}...", WipeLeft, WipeTop)
                If CurrentWindowHeight <> ConsoleWrapper.WindowHeight Or CurrentWindowWidth <> ConsoleWrapper.WindowWidth Then ResizeSyncing = True
                If Not ResizeSyncing Then
                    ConsoleWrapper.SetCursorPosition(WipeLeft, WipeTop)
                    SetConsoleColor(New Color(LighterBackgroundColor), True)
                    WritePlain(" ", False)
                    CoveredPositions.RemoveAt(0)
                Else
                    WdbgConditional(ScreensaverDebug, DebugLevel.W, "Resize-syncing. Clearing covered positions...")
                    CoveredPositions.Clear()
                End If
            End If

            'Reset resize sync
            ResizeSyncing = False
            CurrentWindowWidth = ConsoleWrapper.WindowWidth
            CurrentWindowHeight = ConsoleWrapper.WindowHeight
            SleepNoBlock(LighterDelay, ScreensaverDisplayerThread)
        End Sub

    End Class
End Namespace

File: Kernel Simulator/Misc/Screensaver/Displays/Lighter.cs
﻿//
// Kernel Simulator  Copyright (C) 2018-2024  Aptivi
//
// This file is part of Kernel Simulator
//
// Kernel Simulator is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// Kernel Simulator is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY, without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

using System;
using System.Collections.Generic;
using System.Linq;
using KS.Misc.Reflection;
using KS.Misc.Writers.DebugWriters;
using KS.Misc.Threading;
using Terminaux.Base;
using Terminaux.Colors;

namespace KS.Misc.Screensaver.Displays
{
    /// <summary>
    /// Settings for Lighter
    /// </summary>
    public static class LighterSettings
    {
        private static bool _lighter255Colors;
        private static bool _lighterTrueColor = true;
        private static int _lighterDelay = 100;
        private static int _lighterMaxPositions = 10;
        private static string _lighterBackgroundColor = new Color(ConsoleColor.Black).PlainSequence;
        private static int _lighterMinimumRedColorLevel = 0;
        private static int _lighterMinimumGreenColorLevel = 0;
        private static int _lighterMinimumBlueColorLevel = 0;
        private static int _lighterMinimumColorLevel = 0;
        private static int _lighterMaximumRedColorLevel = 255;
        private static int _lighterMaximumGreenColorLevel = 255;
        private static int _lighterMaximumBlueColorLevel = 255;
        private static int _lighterMaximumColorLevel = 255;

        /// <summary>
        /// [Lighter] Enable 255 color support. Has a higher priority than 16 color support.
        /// </summary>
        public static bool Lighter255Colors
        {
            get
            {
                return _lighter255Colors;
            }
            set
            {
                _lighter255Colors = value;
            }
        }
        /// <summary>
        /// [Lighter] Enable truecolor support. Has a higher priority than 255 color support.
        /// </summary>
        public static bool LighterTrueColor
        {
            get
            {
                return _lighterTrueColor;
            }
            set
            {
                _lighterTrueColor = value;
            }
        }
        /// <summary>
        /// [Lighter] How many milliseconds to wait before making the next write?
        /// </summary>
        public static int LighterDelay
        {
            get
            {
                return _lighterDelay;
            }
            set
            {
                _lighterDelay = value;
            }
        }
        /// <summary>
        /// [Lighter] How many positions to write before starting to blacken them?
        /// </summary>
        public static int LighterMaxPositions
        {
            get
            {
                return _lighterMaxPositions;
            }
            set
            {
                _lighterMaxPositions = value;
            }
        }
        /// <summary>
        /// [Lighter] Screensaver background color
        /// </summary>
        public static string LighterBackgroundColor
        {
            get
            {
                return _lighterBackgroundColor;
            }
            set
            {
                _lighterBackgroundColor = value;
            }
        }
        /// <summary>
        /// [Lighter] The minimum red color level (true color)
        /// </summary>
        public static int LighterMinimumRedColorLevel
        {
            get
            {
                return _lighterMinimumRedColorLevel;
            }
            set
            {
                _lighterMinimumRedColorLevel = value;
            }
        }
        /// <summary>
        /// [Lighter] The minimum green color level (true color)
        /// </summary>
        public static int LighterMinimumGreenColorLevel
        {
            get
            {
                return _lighterMinimumGreenColorLevel;
            }
            set
            {
                _lighterMinimumGreenColorLevel = value;
            }
        }
        /// <summary>
        /// [Lighter] The minimum blue color level (true color)
        /// </summary>
        public static int LighterMinimumBlueColorLevel
        {
            get
            {
                return _lighterMinimumBlueColorLevel;
            }
            set
            {
                _lighterMinimumBlueColorLevel = value;
            }
        }
        /// <summary>
        /// [Lighter] The minimum color level (255 colors or 16 colors)
        /// </summary>
        public static int LighterMinimumColorLevel
        {
            get
            {
                return _lighterMinimumColorLevel;
            }
            set
            {
                _lighterMinimumColorLevel = value;
            }
        }
        /// <summary>
        /// [Lighter] The maximum red color level (true color)
        /// </summary>
        public static int LighterMaximumRedColorLevel
        {
            get
            {
                return _lighterMaximumRedColorLevel;
            }
            set
            {
                _lighterMaximumRedColorLevel = value;
            }
        }
        /// <summary>
        /// [Lighter] The maximum green color level (true color)
        /// </summary>
        public static int LighterMaximumGreenColorLevel
        {
            get
            {
                return _lighterMaximumGreenColorLevel;
            }
            set
            {
                _lighterMaximumGreenColorLevel = value;
            }
        }
        /// <summary>
        /// [Lighter] The maximum blue color level (true color)
        /// </summary>
        public static int LighterMaximumBlueColorLevel
        {
            get
            {
                return _lighterMaximumBlueColorLevel;
            }
            set
            {
                _lighterMaximumBlueColorLevel = value;
            }
        }
        /// <summary>
        /// [Lighter] The maximum color level (255 colors or 16 colors)
        /// </summary>
        public static int LighterMaximumColorLevel
        {
            get
            {
                return _lighterMaximumColorLevel;
            }
            set
            {
                _lighterMaximumColorLevel = value;
            }
        }
    }

    /// <summary>
    /// Display code for Lighter
    /// </summary>
    public class LighterDisplay : BaseScreensaver, IScreensaver
    {

        private readonly List<Tuple<int, int>> CoveredPositions = [];

        /// <inheritdoc/>
        public override string ScreensaverName { get; set; } = "Lighter";

        /// <inheritdoc/>
        public override void ScreensaverPreparation()
        {
            // Variable preparations
            CoveredPositions.Clear();
            ColorTools.LoadBackDry(new Color(LighterSettings.LighterBackgroundColor));
            DebugWriter.Wdbg(DebugLevel.I, "Console geometry: {0}x{1}", ConsoleWrapper.WindowWidth, ConsoleWrapper.WindowHeight);
        }

        /// <inheritdoc/>
        public override void ScreensaverLogic()
        {
            ConsoleWrapper.CursorVisible = false;

            // Select a position
            int Left = RandomDriver.RandomIdx(ConsoleWrapper.WindowWidth);
            int Top = RandomDriver.RandomIdx(ConsoleWrapper.WindowHeight);
            DebugWriter.WdbgConditional(ref Screensaver.ScreensaverDebug, DebugLevel.I, "Selected left and top: {0}, {1}", Left, Top);
            ConsoleWrapper.SetCursorPosition(Left, Top);
            if (!CoveredPositions.Any(t => t.Item1 == Left & t.Item2 == Top))
            {
                DebugWriter.WdbgConditional(ref Screensaver.ScreensaverDebug, DebugLevel.I, "Covering position...");
                CoveredPositions.Add(new Tuple<int, int>(Left, Top));
                DebugWriter.WdbgConditional(ref Screensaver.ScreensaverDebug, DebugLevel.I, "Position covered. Covered positions: {0}", CoveredPositions.Count);
            }

            // Select a color and write the space
            if (LighterSettings.LighterTrueColor)
            {
                int RedColorNum = RandomDriver.Random(LighterSettings.LighterMinimumRedColorLevel, LighterSettings.LighterMaximumRedColorLevel);
                int GreenColorNum = RandomDriver.Random(LighterSettings.LighterMinimumGreenColorLevel, LighterSettings.LighterMaximumGreenColorLevel);
                int BlueColorNum = RandomDriver.Random(LighterSettings.LighterMinimumBlueColorLevel, LighterSettings.LighterMaximumBlueColorLevel);
                DebugWriter.WdbgConditional(ref Screensaver.ScreensaverDebug, DebugLevel.I, "Got color (R;G;B: {0};{1};{2})", RedColorNum, GreenColorNum, BlueColorNum);
                var ColorStorage = new Color(RedColorNum, GreenColorNum, BlueColorNum);
                if (!ConsoleResizeHandler.WasResized(false))
                {
                    ColorTools.SetConsoleColorDry(ColorStorage, true);
                    ConsoleWrapper.Write(" ");
                }
                else
                {
                    DebugWriter.WdbgConditional(ref Screensaver.ScreensaverDebug, DebugLevel.W, "Resize-syncing. Clearing covered positions...");
                    CoveredPositions.Clear();
                }
            }
            else
            {
                int ColorNum = RandomDriver.Random(LighterSettings.LighterMinimumColorLevel, LighterSettings.LighterMaximumColorLevel);
                DebugWriter.WdbgConditional(ref Screensaver.ScreensaverDebug, DebugLevel.I, "Got color ({0})", ColorNum);
                if (!ConsoleResizeHandler.WasResized(false))
                {
                    ColorTools.SetConsoleColorDry(new Color(ColorNum), true);
                    ConsoleWrapper.Write(" ");
                }
                else
                {
                    DebugWriter.WdbgConditional(ref Screensaver.ScreensaverDebug, DebugLevel.W, "Resize-syncing. Clearing covered positions...");
                    CoveredPositions.Clear();
                }
            }

            // Simulate a trail effect
            if (CoveredPositions.Count == LighterSettings.LighterMaxPositions)
            {
                DebugWriter.WdbgConditional(ref Screensaver.ScreensaverDebug, DebugLevel.I, "Covered positions exceeded max positions of {0}", LighterSettings.LighterMaxPositions);
                int WipeLeft = CoveredPositions[0].Item1;
                int WipeTop = CoveredPositions[0].Item2;
                DebugWriter.WdbgConditional(ref Screensaver.ScreensaverDebug, DebugLevel.I, "Wiping in {0}, {1}...", WipeLeft, WipeTop);
                if (!ConsoleResizeHandler.WasResized(false))
                {
                    ConsoleWrapper.SetCursorPosition(WipeLeft, WipeTop);
                    ColorTools.SetConsoleColorDry(new Color(LighterSettings.LighterBackgroundColor), true);
                    ConsoleWrapper.Write(" ");
                    CoveredPositions.RemoveAt(0);
                }
                else
                {
                    DebugWriter.WdbgConditional(ref Screensaver.ScreensaverDebug, DebugLevel.W, "Resize-syncing. Clearing covered positions...");
                    CoveredPositions.Clear();
                }
            }

            // Reset resize sync
            ConsoleResizeHandler.WasResized();
            ThreadManager.SleepNoBlock(LighterSettings.LighterDelay, ScreensaverDisplayer.ScreensaverDisplayerThread);
        }

    }
}